File: CLI Pulse Bar/CLIPulseCore/Sources/CLIPulseCore/TerminalOutputCoalescer.swift
import Foundation

/// Coalesces high-frequency byte chunks into ≤ 1-per-frame batches
/// before handing off to a (typically expensive) sink — addresses
/// Codex HIGH H1 from the v1.24 in-app-terminal plan §Phase 3b.
///
/// Even with the JS-side `requestAnimationFrame` batcher inside
/// xterm.js, the Swift → WKWebView bridge can still see 50–500
/// `evaluateJavaScript` calls per second when the child process
/// produces dense output (think `npm install`, `cargo build`, or a
/// `find /` dump). Each bridge crossing is a JSON marshal + thread
/// hop + JavaScriptCore enter — cheap individually, painful at
/// 500/sec.
///
/// This coalescer adds a 16 ms wall-clock window on the producer
/// side: each `append(_:)` enqueues bytes into a buffer; the first
/// append schedules a flush on the configured queue 16 ms later
/// (one display frame at 60 Hz); subsequent appends within that
/// window merge in without rescheduling. When the timer fires, the
/// buffered concatenation is handed to `onFlush` in a single call.
///
/// At idle frequency (occasional output) the coalescer adds ≤ 16 ms
/// latency to the first byte. At high frequency it caps bridge
/// crossings to ~60/sec regardless of input rate. Combined with the
/// JS-side rAF batcher, the WebView renders one `term.write()` per
/// frame at worst.
///
/// Thread-safety: callers may `append` from any queue; `onFlush`
/// fires on the queue passed at construction (default: a serial
/// dedicated queue). Use `.main` if the consumer must touch the
/// WKWebView directly.
public final class TerminalOutputCoalescer: @unchecked Sendable {

    public typealias FlushHandler = @Sendable (Data) -> Void

    /// Flush window. 16 ms ≈ one frame at 60 Hz; xterm.js renders on
    /// rAF anyway so this matches the JS side's natural cadence.
    public let windowSeconds: TimeInterval

    private let flushQueue: DispatchQueue
    private let onFlush: FlushHandler

    private let lock = NSLock()
    private var pending = Data()
    private var scheduled = false
    /// Test hook: when set, the coalescer asks this closure for
    /// "now" instead of using DispatchTime. Production code uses the
    /// nil branch (real clock); tests inject a virtual clock to
    /// avoid sleeping.
    private let clockProvider: (@Sendable () -> DispatchTime)?

    public init(
        windowSeconds: TimeInterval = 0.016,
        flushQueue: DispatchQueue = DispatchQueue(label: "TerminalOutputCoalescer", qos: .userInitiated),
        onFlush: @escaping FlushHandler,
        clockProvider: (@Sendable () -> DispatchTime)? = nil
    ) {
        precondition(windowSeconds > 0, "windowSeconds must be positive")
        self.windowSeconds = windowSeconds
        self.flushQueue = flushQueue
        self.onFlush = onFlush
        self.clockProvider = clockProvider
    }

    /// Append a chunk. If no flush is scheduled, schedule one
    /// `windowSeconds` from now; otherwise just merge into pending.
    public func append(_ chunk: Data) {
        if chunk.isEmpty { return }
        var shouldSchedule = false
        lock.lock()
        pending.append(chunk)
        if !scheduled {
            scheduled = true
            shouldSchedule = true
        }
        lock.unlock()
        if shouldSchedule {
            let deadline: DispatchTime
            if let clockProvider {
                deadline = clockProvider() + windowSeconds
            } else {
                deadline = .now() + windowSeconds
            }
            flushQueue.asyncAfter(deadline: deadline) { [weak self] in
                self?.flush()
            }
        }
    }

    /// Force-flush whatever's pending right now (synchronous on
    /// the caller's thread, with `onFlush` invoked inline). Useful
    /// when the consuming view unmounts and we want to drop the
    /// timer ASAP without waiting for the deadline.
    public func flushNow() {
        let toSend: Data?
        lock.lock()
        if pending.isEmpty {
            toSend = nil
        } else {
            toSend = pending
            pending = Data()
        }
        scheduled = false
        lock.unlock()
        if let toSend { onFlush(toSend) }
    }

    private func flush() {
        let toSend: Data?
        lock.lock()
        if pending.isEmpty {
            toSend = nil
        } else {
            toSend = pending
            pending = Data()
        }
        scheduled = false
        lock.unlock()
        if let toSend { onFlush(toSend) }
    }

    // MARK: - introspection (test/diagnostic only)

    /// Bytes currently buffered (pre-flush). For tests/diagnostics —
    /// do not use to gate production logic.
    public var pendingByteCount: Int {
        lock.lock(); defer { lock.unlock() }
        return pending.count
    }

    /// True if a flush is currently scheduled.
    public var hasScheduledFlush: Bool {
        lock.lock(); defer { lock.unlock() }
        return scheduled
    }
}

File: android/app/src/main/java/com/clipulse/android/terminal/TerminalOutputCoalescer.kt
package com.clipulse.android.terminal

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.io.ByteArrayOutputStream

/**
 * Schedules a single delayed flush. Injectable so unit tests can fire the
 * flush deterministically instead of sleeping a real 16 ms.
 */
fun interface CoalescerScheduler {
    fun schedule(delayMs: Long, action: () -> Unit)
}

/**
 * Production scheduler: posts the flush to the **main thread** after [delayMs].
 * `WebView.evaluateJavascript` must run on the UI thread, and the flush sink
 * calls it — so the coalesced batch has to land on Main.
 */
class MainCoalescerScheduler(
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main),
) : CoalescerScheduler {
    override fun schedule(delayMs: Long, action: () -> Unit) {
        scope.launch {
            delay(delayMs)
            action()
        }
    }
}

/**
 * v1.27 E4 — Kotlin port of the iOS `TerminalOutputCoalescer`. Merges
 * high-frequency byte chunks into ≤ 1-per-window batches before the expensive
 * `WebView` bridge crossing (`evaluateJavascript("window.pushChunk('…')")`).
 *
 * Even with the JS-side rAF batcher in the bundle, dense PTY output (`npm
 * install`, `cargo build`) can drive 50–500 bridge crossings/sec; each is a
 * JSON marshal + thread hop into the JS engine. A 16 ms wall-clock window
 * (≈ one 60 Hz frame) caps that to ~60/sec regardless of input rate, matching
 * the JS side's natural cadence. (Plan R4.)
 *
 * The first [append] schedules a flush [windowMs] out; subsequent appends
 * within the window merge in without rescheduling. Thread-safe: [append] may
 * be called from any thread (OkHttp's WS dispatcher), [onFlush] fires on the
 * scheduler's thread (Main in production).
 */
class TerminalOutputCoalescer(
    private val windowMs: Long = 16L,
    private val scheduler: CoalescerScheduler = MainCoalescerScheduler(),
    private val onFlush: (ByteArray) -> Unit,
) {
    init {
        require(windowMs > 0) { "windowMs must be positive" }
    }

    private val lock = Any()
    private val pending = ByteArrayOutputStream()
    private var scheduled = false

    /** Queue a chunk; schedule a flush if none is pending. Empty chunks are ignored. */
    fun append(chunk: ByteArray) {
        if (chunk.isEmpty()) return
        val shouldSchedule: Boolean
        synchronized(lock) {
            pending.write(chunk)
            shouldSchedule = !scheduled
            if (shouldSchedule) scheduled = true
        }
        if (shouldSchedule) scheduler.schedule(windowMs) { emit() }
    }

    /**
     * Force-flush whatever is buffered right now, synchronously on the
     * caller's thread. Used when the view tears down so we don't wait out the
     * window. A flush scheduled before this call becomes a no-op (buffer empty).
     */
    fun flushNow() = emit()

    private fun emit() {
        val toSend = synchronized(lock) {
            val batch = if (pending.size() == 0) null else pending.toByteArray()
            pending.reset()
            scheduled = false
            batch
        }
        if (toSend != null) onFlush(toSend)
    }

    /** Bytes currently buffered (pre-flush). Test/diagnostic only. */
    val pendingByteCount: Int get() = synchronized(lock) { pending.size() }

    /** True if a flush is currently scheduled. Test/diagnostic only. */
    val hasScheduledFlush: Boolean get() = synchronized(lock) { scheduled }
}
